src/relay/pass/dispatch_global_func.cc

```diff
@@ -0,0 +1,255 @@
+                         Attrs(take_attrs), {});
+        Array<Expr> cond_list;
+        for (auto const& range : list_ranges) {
+          int64_t low = range[0]->value;
+          int64_t high = range[1]->value;
+          Call cond;
+          CHECK_GT(low, 0)
+            << "Low end of a dispatched interval must be positive.";
+          NDArray low_tensor =
+            NDArray::Empty({1}, DataType::Int(64), cpu_ctx);
+          int64_t* low_tensor_data =
+            reinterpret_cast<int64_t*>(low_tensor->data);
+          low_tensor_data[0] = low;
+          auto cond_attrs = make_node<ReduceAttrs>();
+          Call low_cond =
+            CallNode::make(Op::Get("greater_equal"),
+                           Array<Expr>({dim_val,
+                                        ConstantNode::make(low_tensor)}));
+          if (high > 0) {
+            NDArray high_tensor =
+              NDArray::Empty({1}, DataType::Int(64), cpu_ctx);
+            int64_t* high_tensor_data =
+              reinterpret_cast<int64_t*>(high_tensor->data);
+            high_tensor_data[0] = high;
+            Call high_cond =
+              CallNode::make(Op::Get("less"),
+                             Array<Expr>({dim_val,
+                                          ConstantNode::make(high_tensor)}));
+            cond = CallNode::make(Op::Get("all"),
+                                  Array<Expr>({TupleNode::make(
+                                    Array<Expr>({low_cond, high_cond}))}),
+                                  Attrs(cond_attrs),
+                                  {});
+          } else {
+            cond = CallNode::make(Op::Get("all"),
+                                  Array<Expr>({TupleNode::make(
+                                    Array<Expr>({low_cond}))}),
+                                  Attrs(cond_attrs),
+                                  {});
+          }
+          cond_list.push_back(cond);
+        }
+        conditions.push_back(cond_list);
+      }
+    }
+  }
+
+  // Build dispatching tree
+  // Assume we are dispatching a resnet with dynamic batching input.
+  // We use uniform dispatch function. The higher limit is 256 and
+  // step is 64. Updated main function would be:
+  //
+  //    bs = shape_of(input)[0]
+  //    if 1 <= bs < 64:
+  //        main_copy_0(input)
+  //    elif 64 <= bs < 128:
+  //        main_copy_1(input)
+  //    elif 128 <= bs < 192:
+  //        main_copy_2(input)
+  //    elif 192 <= bs < 256:
+  //        main_copy_3(input)
+  //    else:
+  //        main_copy_4(input)
+  auto new_func_body = DispatchTreeBuilder(mod, func_name, conditions)
+    .BuildDispatchTree(0, 0);
+  mod->Update(mod->GetGlobalVar(func_name),
+              FunctionNode::make(FreeVars(new_func_body),
+              new_func_body,
+              func->ret_type,
+              func->type_params,
+              func->attrs));
+
+  return mod;
+}
+
+TVM_REGISTER_API("relay._transform.add_dispatch_func")
```

Comment:
I think this pass should be implemented as a `ModulePass` so that we don't require users to pass `Module` around. One example is like the following:

https://github.com/apache/incubator-tvm/blob/475158f6285c63b42efe574cb9ba8afec24261be/src/relay/pass/eta_expand.cc#L155-L161